dcpy/connectors/socrata/publish.py

```diff
@@ -102,73 +136,100 @@ def __init__(self, data):
+    A Data Source is what's returned when a dataset is pushed. It contains
+    data about how to modify metadata for the pushed file,
+    e.g. Column Names
+    """
 
-                assert set(self.column_names) == set(
-                    expected_api_names
-                ), f"""The field names in the uploaded data do not match our metadata.
-                - Present in our metadata, but not dataset page: {sorted(set(expected_api_names) - set(self.column_names))}
-                - Present in dataset page, but not our metadata: {sorted(set(self.column_names) - set(expected_api_names))}
-                """
-
-                for uploaded_col in self.soc_output_columns:
-                    # Take the Socrata metadata for columns that have been uploaded,
-                    # modify them to match our metadata, and post it back.
-
-                    # Input columns need to be matched to what's been uploaded,
-                    # via the `initial_output_column_id`. Otherwise update
-                    # requests are ignored.
-                    uploaded_col["initial_output_column_id"] = uploaded_col["id"]
-                    our_col_index = expected_api_names.index(uploaded_col["field_name"])
-                    our_col = dest_col_metadata[our_col_index]
-
-                    uploaded_col["position"] = our_col_index + 1
-
-                    uploaded_col["is_primary_key"] = our_col.is_primary_key
-                    uploaded_col["display_name"] = our_col.display_name
-                    uploaded_col["description"] = our_col.description
-
-                return _socrata_request(
-                    self._column_update_endpoint,
-                    "POST",
-                    json={"output_columns": self.soc_output_columns},
-                )
+    MISSING_COLUMN_ERROR = (
+        "The field names in the uploaded data do not match our metadata"
+    )
+
+    uploaded_columns: list[Any]
+    column_update_endpoint: str
+    _raw_socrata_source: dict[str, Any] | None = None
+
+    @classmethod
+    def from_socrata_source(cls, socrata_source):
+        return RevisionDataSource(
+            _raw_socrata_source=socrata_source,
+            # AR Note: It's not clear to me why/when you would have multiple schemas/output
+            # schemas for any one datasource. In any case, I haven't yet encountered it,
+            # and for our use case (single dataset upload per product) I don't expect we will
+            uploaded_columns=socrata_source.attributes["schemas"][0]["output_schemas"][
+                0
+            ]["output_columns"],
+            column_update_endpoint=f"https://{SOCRATA_DOMAIN}"
+            + socrata_source.input_schemas[0].links["transform"],
+        )
+
+    @property
+    def column_names(self) -> list[str]:
+        return [c["field_name"] for c in self.uploaded_columns]
+
+    def calculate_pushed_col_metadata(self, our_columns: list[md.DatasetColumn]):
+        # TODO: using c.id or c.name?
```

Comment:
I'm not entirely sure how to handle this. Most of the our metadata.column ids are 1-1 with the socrata column, while the column `name` is used for the socrata display name. 

The "correct" way to handle this is to add an "api_name" to the destination overrides, but that just creates a lot of boilerplate. We could potentially add an `external_id` field to the columns, that defaults to the `id` when not present.